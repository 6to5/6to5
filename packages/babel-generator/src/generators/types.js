/* eslint max-len: 0 */
/* eslint quotes: 0 */

import * as t from "babel-types";

export function Identifier(node: Object) {
  // FIXME: We hang variance off Identifer to support Flow's def-site variance.
  // This is a terrible hack, but changing type annotations to use a new,
  // dedicated node would be a breaking change. This should be cleaned up in
  // the next major.
  if (node.variance) {
    if (node.variance === "plus") {
      this.token("+");
    } else if (node.variance === "minus") {
      this.token("-");
    }
  }

  this.word(node.name);
}

export function RestElement(node: Object) {
  this.token("...");
  this.print(node.argument, node);
}

export {
  RestElement as SpreadElement,
  RestElement as SpreadProperty,
  RestElement as RestProperty,
};

export function ObjectExpression(node: Object) {
  let props = node.properties;

  this.token("{");
  this.printInnerComments(node);

  if (props.length) {
    this.space();
    this.printList(props, node, { indent: true, statement: true });
    this.space();
  }

  this.token("}");
}

export { ObjectExpression as ObjectPattern };

export function ObjectMethod(node: Object) {
  this.printJoin(node.decorators, node);
  this._method(node);
}

export function ObjectProperty(node: Object) {
  this.printJoin(node.decorators, node);

  if (node.computed) {
    this.token("[");
    this.print(node.key, node);
    this.token("]");
  } else {
    // print `({ foo: foo = 5 } = {})` as `({ foo = 5 } = {});`
    if (t.isAssignmentPattern(node.value) && t.isIdentifier(node.key) && node.key.name === node.value.left.name) {
      this.print(node.value, node);
      return;
    }

    this.print(node.key, node);

    // shorthand!
    if (node.shorthand &&
      (t.isIdentifier(node.key) &&
       t.isIdentifier(node.value) &&
       node.key.name === node.value.name)) {
      return;
    }
  }

  this.token(":");
  this.space();
  this.print(node.value, node);
}

export function ArrayExpression(node: Object) {
  let elems = node.elements;
  let len   = elems.length;

  this.token("[");
  this.printInnerComments(node);

  for (let i = 0; i < elems.length; i++) {
    let elem = elems[i];
    if (elem) {
      if (i > 0) this.space();
      this.print(elem, node);
      if (i < len - 1) this.token(",");
    } else {
      // If the array expression ends with a hole, that hole
      // will be ignored by the interpreter, but if it ends with
      // two (or more) holes, we need to write out two (or more)
      // commas so that the resulting code is interpreted with
      // both (all) of the holes.
      this.token(",");
    }
  }

  this.token("]");
}

export { ArrayExpression as ArrayPattern };

export function RegExpLiteral(node: Object) {
  this.word(`/${node.pattern}/${node.flags}`);
}

export function BooleanLiteral(node: Object) {
  this.word(node.value ? "true" : "false");
}

export function NullLiteral() {
  this.word("null");
}

export function NumericLiteral(node: Object) {
  let raw = this.getPossibleRaw(node);

  this.number(raw == null ? node.value + "" : raw);
}

export function StringLiteral(node: Object, parent: Object) {
  // raw is the string literal's representation as it occurs
  // in the source code. This preserves quotes, line break
  // characters and unicode code units.
  let val = this.getPossibleRaw(node);
  // fall back to stringified value if raw does not exist
  if (val == null) {
    val = JSON.stringify(node.value);
  }

  // escape illegal js but valid json unicode characters
  val = val.replace(/[\u000A\u000D\u2028\u2029]/g, function (c) {
    return "\\u" + ("0000" + c.charCodeAt(0).toString(16)).slice(-4);
  });

  // Early return when inside JSX, because quotes
  // should not be altered in this case.
  if (t.isJSX(parent)) {
    return this.token(val);
  }

  // remove current quotes (keep originalVal for comparison below)
  let originalVal = val;
  val = val.slice(1, -1);

  if (this.format.quotes === "single") {
    // unescape double quotes
    val = val.replace(/\\"/g, '"');

    // escape single quotes not already escaped
    val = val.replace(/\\([\s\S])|(')/g,"\\$1$2");

    // add single quotes
    val = `'${val}'`;
  }

  // double quotes - default
  if (this.format.quotes !== "single") {
    // unescape single quotes
    val = val.replace(/\\'/g, "'");

    // escape double quotes not already escaped
    val = val.replace(/\\([\s\S])|(")/g,"\\$1$2");

    // add double quotes
    val = `"${val}"`;
  }

  // If conversion to led to escaping quotes and
  // original string had no escaped quotes, prefer
  // the original if avoidEscapingQuotes is true.
  if (this.format.avoidEscapingQuotes
    && /\\'|\\"/.test(val)
    && !/\\'|\\"/.test(originalVal)) {
    return this.token(originalVal);
  }

  return this.token(val);
}
