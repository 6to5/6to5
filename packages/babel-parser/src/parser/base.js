import { plugins } from "./index";

export function hasPlugin(name: string): boolean {
  return plugins.has(name);
}

export function getPluginOption(plugin: string, name: boolean) {
  // $FlowIssue
  if (hasPlugin(plugin)) return plugins.get(plugin)[name];
}
