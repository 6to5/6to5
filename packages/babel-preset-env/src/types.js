//@flow

import { TargetNames, ModulesOption, InjectPolyfillsOption } from "./options";

// Targets
export type Target = $Keys<typeof TargetNames>;
export type Targets = {
  [target: Target]: string,
};

// Options
// Use explicit modules to prevent typo errors.
export type ModuleOption = $Values<typeof ModulesOption>;
export type PolyfillsOption = $Values<typeof InjectPolyfillsOption>;

export type Options = {
  configPath: string,
  debug: boolean,
  exclude: Array<string | RegExp>,
  forceAllTransforms: boolean,
  ignoreBrowserslistConfig: boolean,
  include: Array<string | RegExp>,
  loose: boolean,
  modules: ModuleOption,
  shippedProposals: boolean,
  spec: boolean,
  targets: Targets,
  useBuiltIns: boolean,
  injectPolyfills: PolyfillsOption,
};

// Babel
export type Plugin = [Object, Object];
