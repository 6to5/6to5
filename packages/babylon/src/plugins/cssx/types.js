import { TokenType, types as tt } from "../../tokenizer/types";
import { TokContext, types as tc } from "../../tokenizer/context";

tt.cssxSelector = new TokenType('cssxSelector');
tt.cssxRulesStart = new TokenType('cssxRulesStart');
tt.cssxRulesEnd = new TokenType('cssxRulesEnd');
tt.cssxProperty = new TokenType('cssxProperty');
tt.cssxValue = new TokenType('cssxValue');
tt.cssxMediaQuery = new TokenType('cssxMediaQuery');
tt.cssxMediaQueryStart = new TokenType('cssxMediaQueryStart');
tt.cssxMediaQueryEnd = new TokenType('cssxMediaQueryEnd');

tt.cssxRulesStart.updateContext = function (prevType) {
  if (prevType === tt.cssxSelector) this.state.context.push(tc.cssxRules);
};
tt.cssxRulesEnd.updateContext = function (prevType) {
  if (
    prevType === tt.cssxValue ||
    prevType === tt.cssxRulesStart ||
    prevType === tt.semi
  ) {
    this.state.context.length -= 1; // out of cssxRules
    this.cssxOut();
  }
};
tt.cssxSelector.updateContext = function (prevType) {
  this.state.context.length -= 1;
};