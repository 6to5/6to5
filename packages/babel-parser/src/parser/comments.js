// @flow

/*:: declare var invariant; */

import BaseParser from "./base";
import type { Comment, Node } from "../types";
import * as charCodes from "charcodes";

/**
 * A whitespace token containing comments
 * @typedef CommentWhitespace
 * @type {object}
 * @property {number} start - the start of the whitespace token.
 * @property {number} end - the end of the whitespace token.
 * @property {Array<Comment>} comments - the containing comments
 * @property {Node | null} leadingNode - the immediately preceding AST node of the whitespace token
 * @property {Node | null} trailingNode - the immediately following AST node of the whitespace token
 * @property {Node | null} containerNode - the innermost AST node containing the whitespace
 *                                         with minimal size (|end - start|)
 */
export type CommentWhitespace = {
  start: number,
  end: number,
  comments: Array<Comment>,
  leadingNode: Node | null,
  trailingNode: Node | null,
  containerNode: Node | null,
};
/**
 * Merge comments with node's trailingComments or assign comments to be
 * trailingComments.
 *
 * @param {Node} node
 * @param {Array<Comment>} comments
 */
function setTrailingComments(node: Node, comments: Array<Comment>) {
  if (node.trailingComments === undefined) {
    node.trailingComments = comments;
  } else {
    node.trailingComments.unshift(...comments);
  }
}

/**
 * Merge comments with node's innerComments or assign comments to be
 * innerComments.
 *
 * @param {Node} node
 * @param {Array<Comment>} comments
 */
function setInnerComments(node: Node, comments: Array<Comment>) {
  if (node.innerComments === undefined) {
    node.innerComments = comments;
  } else {
    node.innerComments.unshift(...comments);
  }
}

/**
 * Given node and elements array, if elements has non-null element,
 * merge comments to its trailingComments, otherwise merge comments
 * to node's innerComments
 *
 * @param {Node} node
 * @param {Array<Node>} elements
 * @param {Array<Comment>} comments
 */
function adjustInnerComments(
  node: Node,
  elements: Array<Node>,
  commentWS: CommentWhitespace,
) {
  let lastElement = null;
  let i = elements.length;
  while (lastElement === null && i > 0) {
    lastElement = elements[--i];
  }
  if (lastElement === null || lastElement.start > commentWS.start) {
    setInnerComments(node, commentWS.comments);
  } else {
    setTrailingComments(lastElement, commentWS.comments);
  }
}

/** @class CommentsParser */
export default class CommentsParser extends BaseParser {
  addComment(comment: Comment): void {
    if (this.filename) comment.loc.filename = this.filename;
    this.state.comments.push(comment);
  }

  /**
   * Given a newly created AST node _n_, attach _n_ to a comment whitespace _w_ if applicable
   * {@see {@link CommentWhitespace}}
   *
   * @param {Node} node
   * @returns {void}
   * @memberof CommentsParser
   */
  processComment(node: Node): void {
    const { commentStack } = this.state;
    const commentStackLength = commentStack.length;
    if (commentStackLength === 0) return;
    let i = commentStackLength - 1;
    const lastCommentWS = commentStack[i];

    if (lastCommentWS.start === node.end) {
      lastCommentWS.leadingNode = node;
      i--;
    }

    // invariant: for all 0 <= j <= i, let c = commentStack[j], c must satisfy c.end < node.end
    const nodeStart = node.start;
    for (; i >= 0; i--) {
      const commentWS = commentStack[i];
      const commentEnd = commentWS.end;
      if (commentEnd > nodeStart) {
        // by definition of commentWhiteSpace, this implies commentWS.start > nodeStart
        // so node can be a containerNode candidate
        const { containerNode } = commentWS;
        if (containerNode === null) {
          commentWS.containerNode = node;
        } else {
          // a commentWhitespace is considered _attached_ if
          // 1) its leadingNode or trailingNode, if exists, will not change
          // 2) its containerNode have the minimum node length among all other
          //    nodes containing the commentWhitespace
          // In other words, an attached comment whitespace have fixed associated
          // AST node and thus we can finalize the comment whitespace
          this.finalizeComment(commentWS);
          commentStack.splice(i, 1);
        }
      } else {
        if (commentEnd === nodeStart) {
          commentWS.trailingNode = node;
        }
        break;
      }
    }
  }

  /**
   * Assign the comments of comment whitespaces to related AST nodes.
   * Also adjust innerComments following trailing comma.
   *
   * @memberof CommentsParser
   */
  finalizeComment(commentWS: CommentWhitespace) {
    const { comments } = commentWS;
    if (commentWS.leadingNode !== null || commentWS.trailingNode !== null) {
      if (commentWS.leadingNode !== null) {
        setTrailingComments(commentWS.leadingNode, comments);
      }
      if (commentWS.trailingNode !== null) {
        commentWS.trailingNode.leadingComments = comments;
      }
    } else {
      /*:: invariant(commentWS.containerNode !== null) */
      const { containerNode: node, start: commentStart } = commentWS;
      if (this.input.charCodeAt(commentStart - 1) === charCodes.comma) {
        // If a commentWhitespace follows a comma and the containerNode allows
        // list structures with trailing comma, merge it to the trailingComment
        // of the last non-null list element
        switch (node.type) {
          case "ObjectExpression":
          case "ObjectPattern":
            adjustInnerComments(node, node.properties, commentWS);
            break;
          case "CallExpression":
            adjustInnerComments(node, node.arguments, commentWS);
            break;
          case "ArrayExpression":
          case "ArrayPattern":
            adjustInnerComments(node, node.elements, commentWS);
            break;
          case "ExportNamedDeclaration":
          case "ImportDeclaration":
            adjustInnerComments(node, node.specifiers, commentWS);
            break;
          default: {
            setInnerComments(node, comments);
          }
        }
      } else {
        setInnerComments(node, comments);
      }
    }
  }
}
