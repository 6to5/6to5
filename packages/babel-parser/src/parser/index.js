// @flow

import type { Options } from "../options";
import type { File, Expression, JSXOpeningElement } from "../types";
import type { PluginList } from "../plugin-utils";
import { getOptions } from "../options";
import StatementParser from "./statement";
import { types as tt } from "../tokenizer/types";

export type PluginsMap = {
  [key: string]: { [option: string]: any },
};

export default class Parser extends StatementParser {
  // Forward-declaration so typescript plugin can override jsx plugin
  +jsxParseOpeningElementAfterName: (
    node: JSXOpeningElement,
  ) => JSXOpeningElement;

  constructor(options: ?Options, input: string) {
    options = getOptions(options);
    super(options, input);

    this.options = options;
    this.inModule = this.options.sourceType === "module";
    this.input = input;
    this.plugins = pluginsMap(this.options.plugins);
    this.filename = options.sourceFilename;
  }

  parse(): File {
    const file = this.startNode();
    const program = this.startNode();
    this.nextToken();
    return this.parseTopLevel(file, program);
  }

  // Convenience method to parse an Expression only
  getExpression(): Expression {
    this.nextToken();
    const expr = this.parseExpression();
    if (!this.match(tt.eof)) {
      this.unexpected();
    }
    expr.comments = this.state.comments;
    return expr;
  }
}

function pluginsMap(plugins: PluginList): PluginsMap {
  const pluginMap: PluginsMap = (Object.create(null): Object);
  for (const plugin of plugins) {
    const [name, options = {}] = Array.isArray(plugin) ? plugin : [plugin, {}];
    if (!pluginMap[name]) pluginMap[name] = options || {};
  }
  return pluginMap;
}
