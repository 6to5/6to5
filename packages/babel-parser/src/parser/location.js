// @flow

import { getLineInfo, type Position } from "../util/location";

import { input } from "./index";

// This function is used to raise exceptions on parse errors. It
// takes an offset integer (into the current `input`) to indicate
// the location of the error, attaches the position to the end
// of the error message, and then raises a `SyntaxError` with that
// message.

export function raise(
  pos: number,
  message: string,
  {
    missingPluginNames,
    code,
  }: {
    missingPluginNames?: Array<string>,
    code?: string,
  } = {},
): empty {
  const loc = getLineInfo(input, pos);
  message += ` (${loc.line}:${loc.column})`;
  // $FlowIgnore
  const err: SyntaxError & { pos: number, loc: Position } = new SyntaxError(
    message,
  );
  err.pos = pos;
  err.loc = loc;
  if (missingPluginNames) {
    err.missingPlugin = missingPluginNames;
  }
  if (code !== undefined) {
    err.code = code;
  }
  throw err;
}
