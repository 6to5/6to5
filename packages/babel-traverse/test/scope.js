import traverse, { NodePath } from "../lib";
import { parse } from "@babel/parser";
import * as t from "@babel/types";

function getPath(code, options) {
  const ast =
    typeof code === "string" ? parse(code, options) : createNode(code);
  let path;
  traverse(ast, {
    Program: function(_path) {
      path = _path;
      _path.stop();
    },
  });
  return path;
}

function getIdentifierPath(code) {
  const ast = parse(code);
  let nodePath;
  traverse(ast, {
    Identifier: function(path) {
      nodePath = path;
      path.stop();
    },
  });

  return nodePath;
}

function createNode(node) {
  const ast = t.file(t.program(Array.isArray(node) ? node : [node]));

  // This puts the path into the cache internally
  // We afterwards traverse ast, as we need to start traversing
  // at the File node and not the Program node
  NodePath.get({
    hub: {
      buildError: (_, msg) => new Error(msg),
    },
    parentPath: null,
    parent: ast,
    container: ast,
    key: "program",
  }).setContext();

  return ast;
}

describe("scope", () => {
  describe("binding paths", () => {
    it("function declaration id", function() {
      expect(
        getPath("function foo() {}").scope.getBinding("foo").path.type,
      ).toBe("FunctionDeclaration");
    });

    it("function expression id", function() {
      expect(
        getPath("(function foo() {})")
          .get("body")[0]
          .get("expression")
          .scope.getBinding("foo").path.type,
      ).toBe("FunctionExpression");
    });

    it("function param", function() {
      expect(
        getPath("(function (foo) {})")
          .get("body")[0]
          .get("expression")
          .scope.getBinding("foo").path.type,
      ).toBe("Identifier");
    });

    it("variable declaration", function() {
      expect(getPath("var foo = null;").scope.getBinding("foo").path.type).toBe(
        "VariableDeclarator",
      );
      expect(
        getPath("var { foo } = null;").scope.getBinding("foo").path.type,
      ).toBe("VariableDeclarator");
      expect(
        getPath("var [ foo ] = null;").scope.getBinding("foo").path.type,
      ).toBe("VariableDeclarator");
      expect(
        getPath("var { bar: [ foo ] } = null;").scope.getBinding("foo").path
          .type,
      ).toBe("VariableDeclarator");
    });

    it("declare var", function() {
      expect(
        getPath("declare var foo;", { plugins: ["flow"] }).scope.getBinding(
          "foo",
        ).path.type,
      ).toBe("DeclareVariable");
    });

    it("declare function", function() {
      expect(
        getPath("declare function foo(): void;", {
          plugins: ["flow"],
        }).scope.getBinding("foo").path.type,
      ).toBe("DeclareFunction");
    });

    it("declare module", function() {
      expect(
        getPath("declare module foo {};", {
          plugins: ["flow"],
        }).scope.getBinding("foo").path.type,
      ).toBe("DeclareModule");
    });

    it("declare type alias", function() {
      expect(
        getPath("declare type foo = string;", {
          plugins: ["flow"],
        }).scope.getBinding("foo").path.type,
      ).toBe("DeclareTypeAlias");
    });

    it("declare opaque type", function() {
      expect(
        getPath("declare opaque type foo;", {
          plugins: ["flow"],
        }).scope.getBinding("foo").path.type,
      ).toBe("DeclareOpaqueType");
    });

    it("declare interface", function() {
      expect(
        getPath("declare interface Foo {};", {
          plugins: ["flow"],
        }).scope.getBinding("Foo").path.type,
      ).toBe("DeclareInterface");
    });

    it("type alias", function() {
      expect(
        getPath("type foo = string;", {
          plugins: ["flow"],
        }).scope.getBinding("foo").path.type,
      ).toBe("TypeAlias");
    });

    it("opaque type alias", function() {
      expect(
        getPath("opaque type foo = string;", {
          plugins: ["flow"],
        }).scope.getBinding("foo").path.type,
      ).toBe("OpaqueType");
    });

    it("interface", function() {
      expect(
        getPath("interface Foo {};", {
          plugins: ["flow"],
        }).scope.getBinding("Foo").path.type,
      ).toBe("InterfaceDeclaration");
    });

    it("import type", function() {
      expect(
        getPath("import type {Foo} from 'foo';", {
          plugins: ["flow"],
        }).scope.getBinding("Foo").path.type,
      ).toBe("ImportSpecifier");
    });

    it("variable constantness", function() {
      expect(getPath("var a = 1;").scope.getBinding("a").constant).toBe(true);
      expect(getPath("var a = 1; a = 2;").scope.getBinding("a").constant).toBe(
        false,
      );
      expect(getPath("var a = 1, a = 2;").scope.getBinding("a").constant).toBe(
        false,
      );
      expect(
        getPath("var a = 1; var a = 2;").scope.getBinding("a").constant,
      ).toBe(false);
    });

    it("purity", function() {
      expect(
        getPath("({ x: 1 })")
          .get("body")[0]
          .get("expression")
          .isPure(),
      ).toBeTruthy();
      expect(
        getPath("`${a}`")
          .get("body")[0]
          .get("expression")
          .isPure(),
      ).toBeFalsy();
      expect(
        getPath("let a = 1; `${a}`")
          .get("body")[1]
          .get("expression")
          .isPure(),
      ).toBeTruthy();
      expect(
        getPath("let a = 1; `${a++}`")
          .get("body")[1]
          .get("expression")
          .isPure(),
      ).toBeFalsy();
      expect(
        getPath("tagged`foo`")
          .get("body")[0]
          .get("expression")
          .isPure(),
      ).toBeFalsy();
      expect(
        getPath("String.raw`foo`")
          .get("body")[0]
          .get("expression")
          .isPure(),
      ).toBeTruthy();
    });

    test("label", function() {
      expect(getPath("foo: { }").scope.getBinding("foo")).toBeUndefined();
      expect(getPath("foo: { }").scope.getLabel("foo").type).toBe(
        "LabeledStatement",
      );
      expect(getPath("foo: { }").scope.getLabel("toString")).toBeUndefined();

      expect(
        getPath(
          `
      foo: { }
    `,
        ).scope.generateUid("foo"),
      ).toBe("_foo");
    });

    test("generateUid collision check with labels", function() {
      expect(
        getPath(
          `
      _foo: { }
    `,
        ).scope.generateUid("foo"),
      ).toBe("_foo2");

      expect(
        getPath(
          `
      _foo: { }
      _foo1: { }
      _foo2: { }
    `,
        ).scope.generateUid("foo"),
      ).toBe("_foo3");
    });

    it("reference paths", function() {
      const path = getIdentifierPath("function square(n) { return n * n}");
      const referencePaths = path.context.scope.bindings.n.referencePaths;
      expect(referencePaths).toHaveLength(2);
      expect(referencePaths[0].node.loc.start).toEqual({
        line: 1,
        column: 28,
      });
      expect(referencePaths[1].node.loc.start).toEqual({
        line: 1,
        column: 32,
      });
    });
  });

  describe("duplicate bindings", () => {
    /*
     * These tests do not use the parser as the parser has
     * its own scope tracking and we want to test the scope tracking
     * of traverse here and see if it handles duplicate bindings correctly
     */
    describe("catch", () => {
      // try {} catch (e) { let e; }
      const createTryCatch = function(kind) {
        return t.tryStatement(
          t.blockStatement([]),
          t.catchClause(
            t.identifier("e"),
            t.blockStatement([
              t.variableDeclaration(kind, [
                t.variableDeclarator(t.identifier("e"), t.stringLiteral("1")),
              ]),
            ]),
          ),
        );
      };
      ["let", "const"].forEach(name => {
        it(name, () => {
          const ast = createTryCatch(name);

          expect(() => getPath(ast)).toThrowErrorMatchingSnapshot();
        });
      });

      it("var", () => {
        const ast = createTryCatch("var");

        expect(getPath(ast).node).toMatchSnapshot();
      });
    });

    describe("global", () => {
      ["let", "const"].forEach(name => {
        describe(name, () => {
          it("class", () => {
            const ast = [
              t.variableDeclaration(name, [
                t.variableDeclarator(t.identifier("foo")),
              ]),
              t.classDeclaration(t.identifier("foo"), null, t.classBody([])),
            ];

            expect(() => getPath(ast)).toThrowErrorMatchingSnapshot();
          });
          it("function", () => {
            const ast = [
              t.variableDeclaration(name, [
                t.variableDeclarator(t.identifier("foo")),
              ]),
              t.functionDeclaration(
                t.identifier("foo"),
                [],
                t.blockStatement([]),
              ),
            ];

            expect(() => getPath(ast)).toThrowErrorMatchingSnapshot();
          });
          it("var", () => {
            const ast = [
              t.variableDeclaration(name, [
                t.variableDeclarator(t.identifier("foo")),
              ]),
              t.variableDeclaration("var", [
                t.variableDeclarator(t.identifier("foo")),
              ]),
            ];

            expect(() => getPath(ast)).toThrowErrorMatchingSnapshot();
          });
        });
      });
    });
  });
});
