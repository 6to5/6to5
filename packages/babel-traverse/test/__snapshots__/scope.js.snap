// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scope duplicate bindings catch const 1`] = `"Duplicate declaration \\"e\\""`;

exports[`scope duplicate bindings catch let 1`] = `"Duplicate declaration \\"e\\""`;

exports[`scope duplicate bindings catch var 1`] = `
Object {
  "body": Array [
    Object {
      "block": Object {
        "body": Array [],
        "directives": Array [],
        "type": "BlockStatement",
      },
      "finalizer": null,
      "handler": Object {
        "body": Object {
          "body": Array [
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "e",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "1",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "var",
              "type": "VariableDeclaration",
            },
          ],
          "directives": Array [],
          "type": "BlockStatement",
        },
        "param": Object {
          "name": "e",
          "type": "Identifier",
        },
        "type": "CatchClause",
      },
      "type": "TryStatement",
    },
  ],
  "directives": Array [],
  "interpreter": null,
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`scope duplicate bindings global const class 1`] = `"Duplicate declaration \\"foo\\""`;

exports[`scope duplicate bindings global const function 1`] = `"Duplicate declaration \\"foo\\""`;

exports[`scope duplicate bindings global const var 1`] = `"Duplicate declaration \\"foo\\""`;

exports[`scope duplicate bindings global let class 1`] = `"Duplicate declaration \\"foo\\""`;

exports[`scope duplicate bindings global let function 1`] = `"Duplicate declaration \\"foo\\""`;

exports[`scope duplicate bindings global let var 1`] = `"Duplicate declaration \\"foo\\""`;
